File: strategy/DuckSimulator/cpp/duck.cpp
#include "duck.h"
#include <iostream>


Duck::Duck(FlyBehavior* fb, QuackBehavior* qb) {
	if(fb == nullptr)
		_flyBehavior = new FlyLow();
	else
		_flyBehavior = fb;

	if(qb==nullptr)
		_quackBehavior = new QuackQuietly();
	else
		_quackBehavior = qb;

}

void Duck::fly() const {
	_flyBehavior->fly();
}

void Duck::quack() const {
	_quackBehavior->quack();
}

void Duck::setFlyBehavior(FlyBehavior* fb) {
	_flyBehavior = fb;
}

void Duck::setQuackBehavior(QuackBehavior* qb){
	_quackBehavior = qb;
}

File: strategy/DuckSimulator_V1/python/duck.py
from behaviors import FlyHigh, quackNoisy

class Duck:
	def __init__(self):
		self.flyBehavior = FlyHigh()
		self.quackBehavior = quackNoisy()

	def fly(self):
		self.flyBehavior.fly()

	def quack(self):
		self.quackBehavior.quack()

	def setFlyBehavior(self, behavior):
		self.flyBehavior = behavior()

	def setQuackBehavior(self, behavior):
		self.quackBehavior = behavior()
